### PR TITLE
[5.2.0] Support account locking for TOTP

### PR DESCRIPTION
**Pull request by GANGANI**
## Purpose
Backport fix of https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/73
Related PR: https://github.com/wso2-support/carbon-identity-framework/pull/1440

Writing unit tests is in progress. Unit tests will be added before merging the PR.

### CODE REVIEW

File: component/authenticator/pom.xml [hunk 1]
```diff
@@ -96,6 +96,11 @@
             <version>1.6.5</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.orbit.org.owasp.encoder</groupId>
+            <artifactId>encoder</artifactId>
+            <version>${encoder.wso2.version}</version>
```
Comment (emswbandara) [inline component/authenticator/pom.xml:102]: remove version from here and define in parent pom.xml

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 2]
```diff
@@ -275,44 +282,84 @@ private String getErrorPage(AuthenticationContext context) throws Authentication
 	 * @throws AuthenticationFailedException Authentication process failed for user
 	 */
 	@Override
-	protected void processAuthenticationResponse(HttpServletRequest request,
-	                                             HttpServletResponse response,
-	                                             AuthenticationContext context)
-			throws AuthenticationFailedException {
+	protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
+												 AuthenticationContext context) throws AuthenticationFailedException {
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:286]: formatting issue?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 3]
```diff
@@ -469,4 +516,142 @@ private boolean isValidTokenLocalUser(int token, String username, Authentication
...
+			userStoreManager.setUserClaimValues(IdentityUtil.addDomainToName(authenticatedUser.getUserName(),
+					authenticatedUser.getUserStoreDomain()), updatedClaims, UserCoreConstants.DEFAULT_PROFILE);
+		} catch (UserStoreException e) {
+			if (log.isDebugEnabled()) {
+				log.debug("Error while updating user claims of user: " + authenticatedUser.getUserName(), e);
+			}
+			String errorMessage = "Failed to update user claims for user : " + authenticatedUser.getUserName();
+			throw new AuthenticationFailedException(errorMessage, e);
+		}
+	}
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:656]: Add new line EOF

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 4]
```diff
@@ -275,44 +282,84 @@ private String getErrorPage(AuthenticationContext context) throws Authentication
...
-			throws AuthenticationFailedException {
+	protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
+												 AuthenticationContext context) throws AuthenticationFailedException {
+
 		String token = request.getParameter(TOTPAuthenticatorConstants.TOKEN);
 		String username = context.getProperty("username").toString();
+		validateAccountLockStatusForLocalUser(context, username);
+		if (StringUtils.isBlank(token)) {
+			handleTotpVerificationFail(context);
+			throw new AuthenticationFailedException("Empty TOTP in the request. Authentication Failed for user: " +
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:293]:
```suggestion
			throw new AuthenticationFailedException("Empty TOTP in the request. Authentication failed for user: " +
```

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 5]
```diff
@@ -275,44 +282,84 @@ private String getErrorPage(AuthenticationContext context) throws Authentication
...
+					username);
+		}
+		checkTotpEnabled(context, username);
+		try {
+			int tokenValue = Integer.parseInt(token);
+			if (!isValidTokenLocalUser(tokenValue, username, context)) {
+				handleTotpVerificationFail(context);
+				throw new AuthenticationFailedException("Invalid Token. Authentication failed, user :  " + username);
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:301]:
```suggestion
				throw new AuthenticationFailedException("Invalid Token. Authentication failed for user:  " + username);
```

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 6]
```diff
@@ -275,44 +282,84 @@ private String getErrorPage(AuthenticationContext context) throws Authentication
...
+				AuthenticatedUser authenticatedUser = new AuthenticatedUser();
+				authenticatedUser.setAuthenticatedSubjectIdentifier(username);
+				authenticatedUser.setUserName(
+						UserCoreUtil.removeDomainFromName(MultitenantUtils.getTenantAwareUsername(username)));
+				authenticatedUser.setUserStoreDomain(UserCoreUtil.extractDomainFromName(username));
+				authenticatedUser.setTenantDomain(MultitenantUtils.getTenantDomain(username));
+				context.setSubject(authenticatedUser);
+			} else {
+				context.setSubject(AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(username));
+			}
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:313]: Check formatting.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 7]
```diff
@@ -275,44 +282,84 @@ private String getErrorPage(AuthenticationContext context) throws Authentication
...
+			handleTotpVerificationFail(context);
+			throw new AuthenticationFailedException("TOTP Authentication process failed for user " + username, e);
+		} catch (TOTPException e) {
+			throw new AuthenticationFailedException("TOTP Authentication process failed for user " + username, e);
+		}
+		// It reached here means the authentication was successful.
+		resetTotpFailedAttempts(context);
+	}
+
+	private void checkTotpEnabled(AuthenticationContext context, String username) throws AuthenticationFailedException {
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:324]: Shall we revisit the method name?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 8]
```diff
@@ -275,44 +282,84 @@ private String getErrorPage(AuthenticationContext context) throws Authentication
...
-                        context.getProperty(TOTPAuthenticatorConstants.QR_CODE_CLAIM_URL).toString());
-            }
+			if (context.getProperty(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL) != null) {
+				claims.put(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL,
+						context.getProperty(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL).toString());
+			}
+			if (context.getProperty(TOTPAuthenticatorConstants.QR_CODE_CLAIM_URL) != null) {
+				claims.put(TOTPAuthenticatorConstants.QR_CODE_CLAIM_URL,
+						context.getProperty(TOTPAuthenticatorConstants.QR_CODE_CLAIM_URL).toString());
+			}
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:337]: Check formatting.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 9]
```diff
@@ -469,4 +516,142 @@ private boolean isValidTokenLocalUser(int token, String username, Authentication
 					"TOTPTokenVerifier cannot find the property value for encodingMethod");
 		}
 	}
+	/**
+	 * Execute account lock flow for TOTP verification failures.
+	 *
+	 * @param context Authentication context.
+	 * @throws AuthenticationFailedException Exception on authentication failure.
+	 */
+	private void handleTotpVerificationFail(AuthenticationContext context) throws AuthenticationFailedException {
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:525]:
```suggestion
	private void handleTotpVerificationFailure(AuthenticationContext context) throws AuthenticationFailedException {
```

**Review by ashensw — APPROVED** [review]

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java [hunk 10]
```diff
@@ -275,44 +282,84 @@ private String getErrorPage(AuthenticationContext context) throws Authentication
...
+				authenticatedUser.setTenantDomain(MultitenantUtils.getTenantDomain(username));
+				context.setSubject(authenticatedUser);
+			} else {
+				context.setSubject(AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(username));
+			}
+		} catch (NumberFormatException e) {
+			handleTotpVerificationFail(context);
+			throw new AuthenticationFailedException("TOTP Authentication process failed for user " + username, e);
+		} catch (TOTPException e) {
+			throw new AuthenticationFailedException("TOTP Authentication process failed for user " + username, e);
```
Comment (ashensw) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:318]:
```suggestion
			throw new AuthenticationFailedException("TOTP Authentication process failed for user: " + username, e);
		} catch (TOTPException e) {
			throw new AuthenticationFailedException("TOTP Authentication process failed for user: " + username, e);
```

Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java]: Fixed https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/88/commits/b45cb8cefaab8ad666e280c8eef1bbbe8f001972

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 9 above)
Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java]: Fixed https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/88/commits/b45cb8cefaab8ad666e280c8eef1bbbe8f001972

File: component/authenticator/pom.xml (hunk 1 above)
Comment (GANGANI) [reply component/authenticator/pom.xml]: Fixed b45cb8c

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 4 above)
Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java]: Fixed b45cb8c

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 5 above)
Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java]: Fixed b45cb8c

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 6 above)
Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:312]: Checked

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 7 above)
Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java]: Fixed b45cb8c

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 8 above)
Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:336]: Fixed b45cb8c

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java (hunk 2 above)
Comment (GANGANI) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java:285]: Fixed b45cb8c

**Comment by GANGANI** [comment]
This fix was moved to support branch [1].

[1] https://github.com/wso2-support/identity-outbound-auth-totp/pull/1

**Pull request closed without merging**